### ISSUE TITLE
Make using actions typesafe

### ISSUE THREAD

**Issue by ramnivas** [ochrons/diode#11]
As I tried to move over Diode, one of the errors I faced was passing action **type** as a parameter instead of an action **object**. With actions type expected to be of the `AnyRef` type, there was no help from the compiler. I kept trying to figure out why my handler didn't get invoked.

Perhaps an example (not using Diode, but illustrates the problem, nevertheless) will help (which, btw, happens with Akka and similarly typed systems as well).

```scala
def handle: PartialFunction[AnyRef, Unit] = {
  case Foo(i) => println(s"Received Foo with $i")
}

case class Foo(i: Int)

handle(Foo(5))
Received Foo with 5

scala> handle(Foo) // Ouch... no compiler error
scala.MatchError: Foo (of class Foo$)
  at scala.PartialFunction$$anon$1.apply(PartialFunction.scala:253)
  at scala.PartialFunction$$anon$1.apply(PartialFunction.scala:251)
  at $anonfun$handle$1.applyOrElse(<console>:12)
  at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
  ... 33 elided
```

With Diode, I don't get runtime error either since any unhandled type is eventually silently handled.

The solution is quite simple, I think. Just introduce a marker trait `Action` and change all `AnyRef`s expecting an action to `Action`. Since Diode is very new I don't think that will pose too much burden on people already using it and I doubt that anyone will want to use arbitrary types to signify actions (for example, a `String` type), so asking to extend `Action` for all action type seems like a small price to pay for type safety.

On a related note, it will make reading Diode code easier. Currently, there are too many `AnyRef`s out there and that type doesn't guide in figuring such the code.

**Comment by ochrons** [comment]
This approach was considered initially but I decided to go with `AnyRef` for couple of reasons.

1. Allow actions to be normal, existing case classes in the application.
2. Allow use of special actions such as `Seq[...]` and `None`

As for catching unhandled actions, Diode does have that builtin but you need to supply your own `handleError` function.

```scala
    case action =>
      handleError(s"Action $action was not handled by any action handler")
      ActionResult.NoChange
```

A better way to provide type safety would be to require an implicit typeclass to be present for the action. But even this approach would mean a lot of extra work if the developer just wants to use existing case classes as actions.

**Comment by ramnivas** [comment]
> Allow actions to be normal, existing case classes in the application.

In theory, this sounds okay. But I can't think of significant use cases where using an existing case class will be so essential. One use case, perhaps, is that an existing class already contains structure needed for an action. In that case, wrapping it or implicitly converting to a type that extends `Action` doesn't sound too bad.

> Allow use of special actions such as Seq[...] and None

Could be taken care by `CompositionAction(actions: Seq[Action])` and `case object NoAction extends Action`.

> A better way to provide type safety would be to require an implicit typeclass to be present for the action. But even this approach would mean a lot of extra work if the developer just wants to use existing case classes as actions.

I think you are giving significant importance to the possibility of using existing case classes. Given Diode's (and in fact the Redux/Elm approach's) newness, I doubt its real value.

Elm seems to take the approach of providing compile-time safety for the same scenario. https://github.com/evancz/elm-architecture-tutorial/blob/master/examples/1/Counter.elm#L15. Granted, both defining and dealing with `Action` types is in the application code alone. However, given that Diode want to take things further by providing some additional support in the library, it would have been better, IMO, to not sacrifice on compile-time safety for (in my experience anyway) limited value for supporting any action type.

In the end, it is a judgement call about providing compile-time safety vs. cost of providing such safety.

**Comment by ochrons** [comment]
The scenario I had in my mind is where the server is sending updates to the client and these are directly dispatched. Within the client having an `Action` supertype might be quite feasible, but I wouldn't want to leak client specifics on the server side.

**Comment by ramnivas** [comment]
I, too, wouldn't want to leak client specifics on server side. But I think in this particular use case, asking the client to either wrap specific server types (only those that denote "action") in a class that extends `Action` (or an implicit conversion to one) is how I would approach such integration.

My basic issue is using `AnyRef` to denote actions is too much price to pay for some potential use cases (which are solvable by a wrap-in-an-action approach). Not only this pushes a few compile-time errors to runtime (once someone adds the handler you mentioned), but also doesn't guide in design, code completion, reading Diode code, etc.

In any case, treat it as a feedback based on experience during my switch-to-Diode POC exercise.

**Comment by PerWiklander** [comment]
Can't we just add a type parameter to `Circuit` and `Dispatcher` (and how many other places?) and let the user decide what type they need?

**Comment by ochrons** [comment]
Yesterday I was also thinking about adding a type parameter (or a path-dependent type?) and that might work just fine. Have to experiment with it a bit more. The downside of this approach is that it makes very difficult to build common libs for Diode, if they cannot assume anything about the action type. For example `ActionProcessor`'s have to be aware of the action type.

Another approach is to add secondary methods with more strict typing to the interface.

One reason I don't consider having `AnyRef` as the action type to be so bad, is that Akka actors have been using the same for years, so most Scala devs are familiar with the concept (and JS devs don't care :) )

**Comment by PerWiklander** [comment]
There is a good reason for all the time poured into Akka Typed I guess :-)

**Comment by ochrons** [comment]
Sure, but the new Akka Typed approach is very different from the old one and entirely incompatible with the Diode approach :laughing:

**Comment by PerWiklander** [comment]
I was more into the fact that this stuff is not easy :-)

**Comment by netzwerg** [comment]
I wasted quite some time with the very same silly mistake today (referencing the companion type rather than creating a new action instance). Any new ideas to get more type-safety here?

**Comment by ochrons** [comment]
One cheap solution would be to make `handleError` throw by default, so that the missing action handler would be noticed at runtime. Then add a second check for `Singleton` type actions that are not handled and notify the user of a potential issue with a companion object.

This would quickly reveal the issue, without enforcing any strong typing on dispatch.

**Comment by netzwerg** [comment]
Compile-time safety would be preferable, but your suggestion is better than the status quo. Thank you for reopening this issue!

**Comment by rpiaggio** [comment]
I really like the idea of simplicity: all actions must extend a trait `Action`. In fact, that's what I do internally.

Special cases can be taken care by `ActionSeq(actions: Action*)` and `Nop`.

**Comment by thSoft** [comment]
I also vote for the type parameter approach. Where would it get too complicated?

**Comment by ochrons** [comment]
A typeclass option is to define Dispatcher like this,
```scala
trait ActionType[-A]

trait Dispatcher {
  def dispatch[A <: AnyRef](action: A)(implicit ev: ActionType[A]): Unit

  def apply[A <: AnyRef](action: A)(implicit ev: ActionType[A]) = dispatch(action)
}
```
and then you need to provide _evidence_ that whatever you are dispatching is an action, for example with

```scala
implicit object AnyRefAction extends ActionType[AnyRef]
```
or with
```scala
implicit object MyAction extends ActionType[MyAction]
```
for your own `MyAction` type.

**Comment by ochrons** [comment]
There is now a published 0.6.0-SNAPSHOT with these changes, so it would be great if you guys would try it out. I plan to release 0.6.0 by end of this week, so some testing would be appreciated :)

The docs have been updated (but not published), so you can read them inside the repo.

For an easy transition, `import diode.AnyRefAction._` will make anything work as an action, just like before.

**Comment by rpiaggio** [comment]
Thank you,

Seems to be working fine on my projects.

BTW, I imagine the code

```
trait Action

object Action {
  implicit object actionType extends ActionType[Action]
}
```

will be so common that it might make sense to make it available from Diode as an import.

**Comment by ochrons** [comment]
The common `Action` trait has now been added to Diode, to make developers' life easier :)

**Comment by netzwerg** [comment]
Thank you, works fine!

An unrelated issue I noticed when upgrading to 0.6.0-SNAPSHOT:

This is how I currently connect my component into my circuit:
```scala
case UsersLocation => BimgurCircuit.connect(_.users)(CUserList(_))
```
This did not compile anymore: `missing parameter type for expanded function ((x$3) => CUserList(x$3))`

I could only fix it by extracting the connection to a separate variable:
```scala
case UsersLocation =>
  val connection = BimgurCircuit.connect(_.users)
  connection(CUserList(_))
```
Any hints/tricks?

**Comment by ochrons** [comment]
This is an intended change as you can see in the [updated documentation](https://github.com/ochrons/diode/blob/master/doc/UsageWithReact.md).

The reason for change is related to a problem with excessive remounting in the previous design.

**Comment by netzwerg** [comment]
That makes sense, thank you for the clarification!

**Comment by ochrons** [comment]
This is now resolved in 1.0.0 release
